proposals/XXXX-spirv-resource-representation.md

```diff
@@ -0,0 +1,281 @@
+<!-- {% raw %} -->
+
+# HLSL resources in SPIR-V
+
+*   Proposal: [NNNN](NNNN-spirv-resource-representation.md)
+*   Author(s): [Steven Perron](https://github.com/s-perron)
+*   Status: **Design In Progress**
+
+*During the review process, add the following fields as needed:*
+
+*   PRs: [#114273](https://github.com/llvm/llvm-project/pull/114273),
+    [#111052](https://github.com/llvm/llvm-project/pull/111052),
+    [#111564](https://github.com/llvm/llvm-project/pull/111564),
+    [#115178](https://github.com/llvm/llvm-project/pull/115178)
+
+## Introduction
+
+There is a need to represent the HLSL resources in llvm-ir in a way that the
+SPIR-V backend is able to create the correct code. We have already done some
+implementation work for `Buffer` and `RWBuffer`. This was done as a
+proof-of-concept, and now we needed to determine how the other resource types
+will be represented.
+
+## Motivation
+
+The HLSL resources are fundamental to HLSL, and they are required in a Vulkan
+implementation.
+
+## Proposed solution
+
+We want to match the general solution proposed in
+[0006-resource-representations.md](0006-resource-representations.md). The
+`@llvm.spv.handle.fromBinding` intrinsic will be used to get a handle to the
+resource. It will return a target type to represent the handle. Then other
+intrinsics will be used to access the resource using the handle. Previous
+proposals left open what the target types should be for SPIR-V.
+
+The type for the handle will depend on the type of resource, and will be
+detailed in the following sections.
+
+The following sections will reference table 4 in the
+[shader resrouce interface](https://docs.vulkan.org/spec/latest/chapters/interfaces.html#interfaces-resources)
+for Vulkan.
+
+### SPIR-V target types
+
+The must be appropriate SPIR-V target types to represent the HLSL resources. We
+could try to represent the resources using the exact SPIR-V type that will be.
+The problem is that the HLSL resources does not map too closely with SPIR-V.
+
+Consider `StructuredBuffer`, `RWStructuredBuffer`,
+`RasterizerOrderedStructuredBuffer`, `AppendStructureBuffer`, and
+`ConsumeStructuredBuffer`. These resource types do not map directly to SPIR-V.
+They have multiple implicit features that need to map to different SPIR-V:
+
+1.  They all contains an array of memory that maps to a storage buffer.
+2.  Other than `StructuredBuffer`, they all contains a separate counter variable
+    that is its own storage buffer.
+3.  The references to `RasterizerOrderedStructuredBuffer` are contained in
+    implicit critical regions. In SPIR-V, explicit instructions are used to
+    start and stop the critical region.
+
+This makes it impossible to create a handle type that maps directly to a SPIR-V
+type. To handle this, we will create a target type `spv.VulkanBuffer`:
```

Comment:
Would it be better to use a name different than `VulkanBuffer`, since it represents concepts that aren't from Vulkan? (ie, Vulkan doesn't have an exact concept of a rasterizer order view)

Reply:
Also, should it be `spirv.VulkanBuffer` instead, to be more consistent with `spirv.Image`, `spirv.Type`, and so on?

Reply:
@Keenuts @cassiebeckley Do you have any suggestions for the name. At the heart, this encapsulates uniform buffers and storage buffer in vulkan. However, it does include more: rov and the counter variable. I'm not sure how to name it.

Reply:
I would be tempted to make this a language intrinsic instead of `spirv`, something like `hlsl.Buffer` until the second stage of the lowering where you generate the `spv.resource.handleFromBinding` or `spv.resource.UpdateCounter`.

Last time we discussed something similar was the target builtins (https://discourse.llvm.org/t/rfc-add-targetbuiltins-for-spirv-to-support-hlsl/83329)
But back then it was for simple intrinsics which would make sense as target intrinsics and could be reusable across frontends.

But here, since this carries the optional counter & buffer, and has a very hlsl-specific semantic, I feel it would be a good reason to keep it as a language-level builtin.